### ISSUE TITLE
About the performance in disjoint Pascal-Voc (19-1) task.

### ISSUE THREAD

**Issue by CS-GangXu** [fcdl94/MiB#17]
Hi, thank you for creating such an innovative and wonderful incremental semantic segmentation method.
In this paper, the mean IoU of MiB in disjoint Pascal-Voc (19-1) Task is **69.6(1-19) 25.6(20) 67.4(all)**. When I try to reproduce the performance, the final mean IoU of MiB(step1) in disjoint Pascal-Voc (19-1) task is **62.7(1-19) 17.3(20) 60.5(all)**, thus I want to know if there is a problem of my MiB(step0). Could you please show me the mean IoU of your MiB(step0) in disjoint Pascal-Voc (19-1) task?

Here is the test result when my MiB(step0) was trained after 30 epochs:
> Train set: 10034, Val set: 1421, Test set: 1421, n_classes 20
INFO:rank0: *** Test the model on all seen classes...
INFO:rank0: *** Model restored from checkpoints/step/19-1-voc_MIB_original_0.pth
Total samples: 1422.000000
Overall Acc: 0.937206
Mean Acc: 0.850237
FreqW Acc: 0.888375
Mean IoU: 0.753881
Class IoU:
        class 0: 0.9282959437197884
        class 1: 0.8882853094595112
        class 2: 0.3797590436288108
        class 3: 0.8807409395997082
        class 4: 0.6648712821388697
        class 5: 0.7976358668707264
        class 6: 0.9164809865812081
        class 7: 0.8817972918369146
        class 8: 0.9234520147664763
        class 9: 0.32377577234982474
        class 10: 0.8422507677142984
        class 11: 0.5292598617495181
        class 12: 0.8956875294271032
        class 13: 0.8407520660676776
        class 14: 0.848592612779058
        class 15: 0.8356916493750883
        class 16: 0.5890588805790536
        class 17: 0.8099400266733182
        class 18: 0.4580267122561858
        class 19: 0.8432569294818675
Class Acc:
        class 0: 0.966766716830572
        class 1: 0.960997648920115
        class 2: 0.8587795396354625
        class 3: 0.9328072965429103
        class 4: 0.8608227242951523
        class 5: 0.9055918146656152
        class 6: 0.9412850562654307
        class 7: 0.9158580847049574
        class 8: 0.9614393245122251
        class 9: 0.47560739230891674
        class 10: 0.9135487146581773
        class 11: 0.5875200166947372
        class 12: 0.9519971945124238
        class 13: 0.9047293200139962
        class 14: 0.9272278961100133
        class 15: 0.8997701656461145
        class 16: 0.7169984272644204
        class 17: 0.867543489475999
        class 18: 0.5452584289443756
        class 19: 0.9101945156506736

And the test result when my MiB(step1) was trained after 30 epochs:
> Train set: 548, Val set: 74, Test set: 1449, n_classes 21
INFO:rank0: *** Test the model on all seen classes...
INFO:rank0: *** Model restored from checkpoints/step/19-1-voc_MIB_original_1.pth
Total samples: 1450.000000
Overall Acc: 0.890334
Mean Acc: 0.725162
FreqW Acc: 0.820754
Mean IoU: 0.618920  # 0.6411918301227864 0-19 20 0.1734
Class IoU:
        class 0: 0.8938328066648968
        class 1: 0.7782195086558429
        class 2: 0.37033196721590445
        class 3: 0.8413915510994053
        class 4: 0.525415475349433
        class 5: 0.6729141125705496
        class 6: 0.20963013074814107
        class 7: 0.7363044189891342
        class 8: 0.8777653860912165
        class 9: 0.29350207126893696
        class 10: 0.8145631162334964
        class 11: 0.49090401230842257
        class 12: 0.8398493348474195
        class 13: 0.7777190817231865
        class 14: 0.683588348834814
        class 15: 0.7999431095131695
        class 16: 0.555339335033059
        class 17: 0.7869103135756841
        class 18: 0.4080286080402448
        class 19: 0.4676839136927719
        class 20: 0.17347901289607445
Class Acc:
        class 0: 0.9573831275998825
        class 1: 0.7941349223562392
        class 2: 0.7405488472008162
        class 3: 0.939724801161348
        class 4: 0.5830666957769701
        class 5: 0.7291240593393206
        class 6: 0.21006324595180248
        class 7: 0.7776654179327952
        class 8: 0.9586466681798896
        class 9: 0.40889428771720887
        class 10: 0.8555168166200787
        class 11: 0.5290773778328225
        class 12: 0.9181100424143835
        class 13: 0.8976137666936727
        class 14: 0.7100771395812419
        class 15: 0.8855347468407907
        class 16: 0.702358914737102
        class 17: 0.8942199044152891
        class 18: 0.4998755740006467
        class 19: 0.47593462075060633
        class 20: 0.7608375943271062

I run the following commands to train the MiB in disjoint Pascal-Voc (19-1) task.
> python -m torch.distributed.launch --nproc_per_node=2 run.py --data_root data --batch_size 12 --dataset voc --name MIB_original --task 19-1 --step 0 --lr 0.001 --epochs 30 --method MiB
> python -m torch.distributed.launch --nproc_per_node=2 run.py --data_root data --batch_size 12 --dataset voc --name MIB_original --task 19-1 --step 1 --lr 0.001 --epochs 30 --method MiB

**Comment by ygjwd12345** [comment]
as  paper introduce "We use an initial learning rate of 10−2 for the first learning step and 10−3 for the followings, as in [32].", I think  you need to change  the lr value in step 0

**Comment by fcdl94** [comment]
Thanks for the comment! 
The answer of @ygjwd12345 is correct, the learning rate of the first step should be 10−2 and not 10-3.

**Comment by CS-GangXu** [comment]
Thanks for your comments! @fcdl94  @ygjwd12345 
I tried to set the learning rate to 10−2 and 10-3 in the first step and second step, now I got the performance of step1 in VOC 19-1 disjoint task:
> Train set: 548, Val set: 74, Test set: 1449, n_classes 21
INFO:rank0: *** Test the model on all seen classes...
INFO:rank0: *** Model restored from checkpoints/step/19-1-voc_MIB_original_1.pth
Total samples: 1450.000000
Overall Acc: 0.908616
Mean Acc: 0.763308
FreqW Acc: 0.848228
Mean IoU: 0.669702
Class IoU:
        class 0: 0.9025151038845602
        class 1: 0.6770667838145736
        class 2: 0.3859078617512791
        class 3: 0.8607920125480194
        class 4: 0.44872201326348293
        class 5: 0.6467189345698922
        class 6: 0.6986792740621603
        class 7: 0.7787412596170391
        class 8: 0.9116445289975935
        class 9: 0.3697605044528561
        class 10: 0.8181404936094833
        class 11: 0.5181223740450368
        class 12: 0.8634136452097644
        class 13: 0.8252743728819917
        class 14: 0.7407675365756416
        class 15: 0.8366811607862535
        class 16: 0.5823912233470885
        class 17: 0.8598172373323497
        class 18: 0.4882388358081148
        class 19: 0.6383841774191465
        class 20: 0.21196177524410226

and when I changed the learning rate of step1 to 0.0005, the performance is better:
> Train set: 548, Val set: 74, Test set: 1449, n_classes 21
INFO:rank0: *** Test the model on all seen classes...
INFO:rank0: *** Model restored from checkpoints/step/19-1-voc_MIB_original_1.pth
Total samples: 1450.000000
Overall Acc: 0.917885
Mean Acc: 0.773767
FreqW Acc: 0.859465
Mean IoU: 0.687735
Class IoU:
        class 0: 0.9109401367340706
        class 1: 0.7502150651572691
        class 2: 0.38908337088764705
        class 3: 0.8556551822790331
        class 4: 0.48544809776601655
        class 5: 0.6493220137145206
        class 6: 0.7883828190869892
        class 7: 0.8088601641462041
        class 8: 0.9135836922323233
        class 9: 0.3684814258064996
        class 10: 0.8267116290777302
        class 11: 0.5137530193134434
        class 12: 0.8602261008661763
        class 13: 0.8277710526589481
        class 14: 0.7718229665236108
        class 15: 0.8351650716919111
        class 16: 0.5909755693233069
        class 17: 0.8635433124835872
        class 18: 0.4852307550123502
        class 19: 0.7125546951553077
        class 20: 0.23470773145406507

**Comment by fcdl94** [comment]
Nice to hear and thanks for the hint!

**Comment by CS-GangXu** [comment]
The performace in 1-19 is 69.9 which is higher than the performance(69.6) in the paper, and can the IoU of class 20(23.47) be regarded as the random fluctuation(compared with the 25.6 in the paper)? I would be very grateful if you can reply to my confusion.

**Comment by fcdl94** [comment]
Yes, I reimplemented it using different random seeds and it can be possible to obtain slightly different results. However they should be close enough, as it is in your case.
Let me add that the 19-1 is the setting where results are more variable, while 15-5 and on ADE the results are more stable.

**Comment by CS-GangXu** [comment]
Thanks for your comment!
